tt_metal/tt_metal.cpp

```diff
@@ -830,6 +830,23 @@ void DeallocateBuffer(Buffer *buffer) {
     allocator::deallocate_buffer(*buffer->device()->allocator_, buffer);
 }
 
+void SynchronizeWorkerThreads(const std::vector<Device*>& workers) {
+    if (tt::tt_metal::detail::InWorkerThread()) {
+        // Early exit if in a worker thread, since waiting for the worker
+        // queue to become empty inside a worker thread leads to a deadlock
+        // Synchronizing in a worker thread should be a nop by definition
+        return;
+    }
```

Comment:
this is the new change? why does this function get called in the worker thread at all?

Reply:
This is a degenerate case when a user chooses to call a blocking function such as `cpu` inside an op (which runs in a worker thread). This is what happened when the cases requiring host fallback increased for reshape (the host fallback path called blocking `cpu` in the worker which lead to the deadlock).